_episodes/l4_04_parameterized_tests.md

```diff
@@ -0,0 +1,353 @@
+[ RUN      ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0
+[       OK ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0 (0 ms)
+[ RUN      ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/1
+[       OK ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/1 (0 ms)
+[----------] 2 tests from NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 2 tests from 1 test suite ran. (0 ms total)
+[  PASSED  ] 2 tests.
+```
+
+In this output, there are two things worth noting:-
+
+1. As expected, we are now running two tests as compared to just one in case of a for loop.
+2. The test name `NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0` is a combination of the following:-
+    - A Prefix `NetBonusIsCorrectForDifferentYears` coming from INSTANTIATE_TEST_SUITE_P.
+    - Parameterised class name `EmployeeTestParameterised` coming from the first argument of `TEST_P` macro.
+    - Test name `NetBonusIsCorrectForDifferentYears` coming from the second argument of `TEST_P` macro.
+    - Finally, the iteration number.
+
+With this parameterised test, we were able to solve the issues that we were discussing above. However, in doing so, we changed the test fixture and converted it to use `TEST_P` macro. Our previous tests based on `TEST_F` macro will not work anymore as it has been replaced. The important question is: What shall we do so that we can still keep all our useful tests from test fixtures while still being able to add parameterised test. The solution is to combine test fixture with parameterised test and the next subsection explains that.
+
+## 3. Parameterised test based on test fixture
+
+In order to create a parameterised test from a test fixture, all we need to do is to create a parameterised test class which derives from both the test fixture class and `testing::WithParamInterface<T>` class (defined in GoogleTest) to create parameterised tests.
+
+```cpp
+// create a parameterised test class from the fixture defined above.
+class YourParameterisedClass : public YourFixtureClass, 
+                               public WithParamInterface<T> {
+};
+```
+
+For the demonstration purpose, let us assume that we now want to check our tax calculation function `getTaxAmount()` which has more branches as compared to bonus calculation. For complete code, see the file [4_param_test_based_fixture.cpp](../code/Chapter4/4_param_test_based_fixture.cpp). We give a small section of code below for reference.
+
+```cpp
+// Create a test fixture.
+class EmployeeTestFixture : public::testing::Test {
+    public:
+        Employee employee{"John", 25, 8000, 5, 1000};
+
+};
+
+// Test if we can set the name of an employee.
```

Comment:
This code section is very big. The concern is that it might be difficult and slow to walk though and explain each section of the code by voice to the seminar. Could this be made a bit more concise? Maybe remove some of the repeated boilerplate stuff. A deeper dive into the code here could be a great exercise for the viewer.

Reply:
I agree with James. It is useful to illustrate a point with small snippets of code, but big chunks of code just make the lesson unreadable. At the very least, remove the first 3 tests (the `TEST_F` ones) as they are not relevant for the current discussion. Also, it might help to break the big chunk of code into one chunk per class/function and take the comments out of the code and into the lesson, in bullet points, something like:

```markdown
1. Create a `EmployeeTestFixture` as done in the previous episode.
2. Create a structure that holds the input and output values. This structure is used to inject values into the test.

```cpp
struct TestValues{
    double inp_salary;
    double inp_bonus;
    double inp_years_employed;
    double out_tax;
    
    TestValues(double salary, double bonus, double years_employed, double tax) 
              : inp_salary(salary), 
                inp_bonus(bonus), 
                inp_years_employed(years_employed), 
                out_tax(tax) {}
};
```

3. Create a parameterised test class from the fixture.

```cpp
class EmployeeTestParameterisedFixture : public EmployeeTestFixture, 
                                         public WithParamInterface<TestValues> {
};
```
```

etc...

Reply:
@jamesturner246 @dalonsoa 
I agree that the code might be long but I cannot reduce it much. Also, please note
1. We have already discussed most of the concepts in the previous chapter. We do not need to explain each and everything in detail here.
2. I would like to see at least the basic entities that explain how to write parameterised test with fixtures. Most of the code just does that. Extra code has been removed in 02b7b5b7c88133b4f08fd969bb0ff589521b34f9.
3. Even if I separate it into multiple points (as @dalonsoa  mentioned) , it is not going to make code/text shorter. I will not follow this approach. This kind of explanation is useful when we are introducing for the first time as was done in a previous section.  This section is based on the concepts already explained. So, it should be alright to have it one code block (although the overall length has been reduced now).

Reply:
Ok, point taken. But you still have a `TEST_F` here that is not relevant for this section.